### PR TITLE
JS: Extends CredentialsNode class mostly related to JWT authentication packages

### PR DESCRIPTION
**Pull request by am0o0**

### CODE REVIEW

File: javascript/ql/lib/semmle/javascript/security/dataflow/HardcodedCredentialsCustomizations.qll [hunk 1]
```diff
@@ -32,6 +32,16 @@ module HardcodedCredentials {
     ConstantStringSource() { not astNode.getStringValue() = "" }
   }
 
+  class NonProductionFiles extends Sanitizer {
+    NonProductionFiles() {
+      this.getFile()
+          .getLocation()
+          .hasLocationInfo(any(string s |
+              s.regexpMatch(["/.*test[.].*", "/.*demo[.].*", "/.*example[.].*", "/.*sample[.].*"])
```
Comment (p-) [inline javascript/ql/lib/semmle/javascript/security/dataflow/HardcodedCredentialsCustomizations.qll:40]: I'm not sure if this should be added as a sanitizer on this level.

Comment (p-) [reply javascript/ql/lib/semmle/javascript/security/dataflow/HardcodedCredentialsCustomizations.qll:40]: Let's wait for the CodeQL team to review it. But I fear if it adds a lot of FP's without this sanitizer it might not be perfectly suited as addition for this prod query.
On the other hand: excluding test files for less FP's might help to reduce FP's for the `hardcoded-credentials` query as a whole.

**Review by erik-krogh — COMMENTED** [review]
The idea seems OK. 
I haven't looked through the models yet, but I did run a small internal evaluation on them, and that looked OK.  

But first there is a problem with the sanitizer.

I'll take a closer look at the models once that's resolved.

**Comment by erik-krogh** [comment]
> I only included the new where condition on sinks because the constant hardcoded creds can be loaded from a test or example directory. However, the sink using these hardcoded creds should not be in a test or example directory.

Can you try to use `ClassifyFiles.qll` instead? I know those heuristics probably don't match exactly what you're looking for, but those can be improved later.  
(I very much suggest that you don't look into that now, but do that in a later PR if needed, just to keep the scope of this PR low).

**Review by erik-krogh — COMMENTED** [review]
> @erik-krogh if the changes are not good and it is better to move my changes to experimental please let me know. 

No, I think that is fine to get it in the standard suite.  
But I did discuss internally, and agreed that you have some changes to make (beyond the comments below).  

Somehow the `isTestFile` filter should only apply to the JWT keys you've added. 
Probably by creating a new kind for `getCredentialsKind`, and then applying the filter only for that credentials-kind.

File: javascript/ql/src/Security/CWE-798/HardcodedCredentials.ql [hunk 2]
```diff
@@ -16,30 +16,50 @@
...
+      sink.getNode().(Sink).(DefaultCredentialsSink).getKind() = "jwt key" and
+      not source.getNode().asExpr() instanceof ConstantString
+    then not isTestFile(sink.getNode().getFile()) and value = "This hard-coded value"
+    else
+      // sink kind is not "jwt key" and source is  constant string
+      if
+        not sink.getNode().(Sink).(DefaultCredentialsSink).getKind() = "jwt key" and
+        source.getNode().asExpr() instanceof ConstantString
+      then updateMessageWithSourceValue(value, source.getNode(), sink.getNode())
+      else value = "This hard-coded value"
```
Comment (erik-krogh) [inline javascript/ql/src/Security/CWE-798/HardcodedCredentials.ql:63]: Can you revert these changes, and then instead put the filter on the `Sink` class instead?

Comment (erik-krogh) [reply javascript/ql/src/Security/CWE-798/HardcodedCredentials.ql:63]: I'll have a look this week. I just came back from a few weeks of vacation.

Comment (erik-krogh) [reply javascript/ql/src/Security/CWE-798/HardcodedCredentials.ql:63]: You can e.g. do it like this: https://github.com/erik-krogh/ql/commit/38e5c90d2719d090a9c42c46bbf4e3f8f9c6ab17 
(I tried to push directly to your PR, but I can't do that).  

There should probably also be a comment on the new characteristic predicate in `DefaultCredentialsSink`.

**Comment by erik-krogh** [comment]
The autoformatter checks are failing on `javascript/ql/src/Security/CWE-798/HardcodedCredentials.ql` and `javascript/ql/lib/semmle/javascript/security/dataflow/HardcodedCredentialsCustomizations.qll`.  
You can use `codeql query format -i <paths-to-files>` to run the autoformatter.

**Comment by erik-krogh** [comment]
There is still `javascript/ql/src/Security/CWE-798/HardcodedCredentials.ql`, it's still failing the autoformatter checks.

File: javascript/ql/lib/semmle/javascript/security/dataflow/HardcodedCredentialsQuery.qll [hunk 3]
```diff
@@ -35,5 +35,44 @@ class Configuration extends DataFlow::Configuration {
...
+      src = n.getParameter(0).getMember(["x", "y", "n"]).asSink() and
+      trg = n.getReturn().getPromised().asSource()
+    )
+    or
+    exists(DataFlow::CallNode n |
+      n = DataFlow::globalVarRef("TextEncoder").getAnInstantiation().getAMemberCall("encode")
+    |
+      src = n.getArgument(0) and
+      trg = n and
+      n.getLocation().getFile().getRelativePath().matches("%HardcodedCredentials.js%")
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/security/dataflow/HardcodedCredentialsQuery.qll:63]: This line looks like some debug code you forgot to remove?

**Review by erik-krogh — COMMENTED** [review]
There is a single debug-line it looks like you forgot to remove.  
But otherwise I think we're ready to merge 🎉

**Review by erik-krogh — APPROVED** [review]
